### ISSUE TITLE
Using Array<T> instead of T[] results in error: Type must have a '[Symbol.iterator]()' method that returns an iterator.

### ISSUE THREAD

**Issue by arendjr** [Microsoft/TypeScript#17888]
**TypeScript Version:**  2.4.2

**Code**

If my understanding is correct, the three samples below should all be equal, but the last one fails:

```ts
// OK:
for (const a of ["a", "b"]) {
    console.log(a);
}

// Explicit cast to `string[]` still OK:
for (const a of (["a", "b"] as string[])) {
    console.log(a);
}

// Cast to `Array<string>` yields: "Type must have a '[Symbol.iterator]()' method that returns an iterator."
for (const a of (["a", "b"] as Array<string>)) {
    console.log(a);
}
```

**Expected behavior:**

All of the snippets should be OK.

**Actual behavior:**

The third snippet shows the following error in the VS Code tooltip:

    [ts] Type must have a '[Symbol.iterator]()' method that returns an iterator.
    interface Array<T>

**tsconfig.json:**

```json
{
    "compilerOptions": {
        "target": "es2017",
        "module": "commonjs",
        "outDir": "dist",
        "sourceMap": true,
        "strictNullChecks": true,
        "typeRoots": ["./node_modules/@types"],
        "lib": ["es2015", "es2016", "es2017"]
    },
    "include": ["src/**/*.ts"],
    "exclude": ["node_modules"]
}
``

**Comment by aluanhaddad** [comment]
I cannot reproduce this error.

Unrelated FYI, you only need to specify `"lib": ["es2017"]` since that transitively includes the `"es2016"` and so on.

**Comment by kitsonk** [comment]
Well, also `target` of `es2017` automatically includes those libs anyways, the `lib` is wholly redundant.

**Comment by arendjr** [comment]
Interestingly, now that I've logged out and back in, I cannot reproduce it anymore either... Could it be it was an issue in older versions, one that perhaps still was loaded by VS Code in my previous session? Anyway, I'll close it now.

**Comment by aluanhaddad** [comment]
In general it's a good idea to restart your IDE, or try the command line compiler, before concluding that something is working incorrectly. I have to restart VS Code fairly often actually, in spite of the new changes that allow me to switch TypeScript versions on the fly, changes to certain properties of tsconfig.json files, having to do with module resolution for example, sometimes require a restart before vs code will pick them up.

This has improved a lot recently but I think there are still edge cases.

**Comment by medington** [comment]
I ended up getting to this page searching on the error message for this issue. In my case, I had copied a `tsconfig.json` entry from the `create-react-app` and started seeing this error. After a bit of digging I realized this was because that project uses only:

```    "lib": ["es6", "dom"]```

whereas when no entry is used the default value for `lib` with `es6` target would be:

```    "lib": ["es2016", "dom", "dom.iterable", "ScriptHost"] ```

It's the `dom.iterable` entry that does the magic here to avoid the error.

This is documented on the help for the `--lib` option in the [compiler-options reference](http://www.typescriptlang.org/docs/handbook/compiler-options.html).
